File: apps/macos/Sources/CornerTasks/Sync/SyncEvent.swift
import CryptoKit
import Foundation

/// One mutation in transit per `docs/sync-protocol.md` §3. Cleartext fields plus the
/// AES-256-GCM ciphertext+nonce of the encrypted payload.
struct SyncEvent: Codable, Equatable {
    let accountDid: String
    let deviceId: String
    let eventId: String
    let taskId: String
    let updatedAt: String
    let op: String
    let ciphertext: String
    let nonce: String
}

/// Plaintext payload encrypted into `SyncEvent.ciphertext` for `op="upsert"`.
struct EventPlaintext: Equatable {
    var title: String
    var createdAt: Date
    var completedAt: Date?
    var dueDate: Date?
    var order: Int
}

enum SyncEventCodec {
    /// `JSON.stringify` with the fixed key order specified by `docs/sync-protocol.md` §4
    /// (title, createdAt, completedAt, dueDate, order). Hand-built so encoder ordering is
    /// independent of dictionary iteration.
    static func encodePlaintext(_ pt: EventPlaintext) -> Data {
        var s = "{"
        s += "\"title\":" + jsonString(pt.title) + ","
        s += "\"createdAt\":" + jsonString(ISO8601.format(pt.createdAt)) + ","
        s += "\"completedAt\":"
        if let c = pt.completedAt { s += jsonString(ISO8601.format(c)) } else { s += "null" }
        s += ","
        s += "\"dueDate\":"
        if let d = pt.dueDate { s += jsonString(ISO8601.format(d)) } else { s += "null" }
        s += ","
        s += "\"order\":\(pt.order)"
        s += "}"
        return Data(s.utf8)
    }

    static func decodePlaintext(_ data: Data) -> EventPlaintext? {
        guard let any = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        guard let title = any["title"] as? String else { return nil }
        guard let createdStr = any["createdAt"] as? String, let createdAt = ISO8601.parse(createdStr) else { return nil }
        var completedAt: Date? = nil
        if let s = any["completedAt"] as? String { completedAt = ISO8601.parse(s) }
        var dueDate: Date? = nil
        if let s = any["dueDate"] as? String { dueDate = ISO8601.parse(s) }
        let order = (any["order"] as? Int) ?? Int((any["order"] as? NSNumber)?.intValue ?? 0)
        return EventPlaintext(title: title, createdAt: createdAt, completedAt: completedAt, dueDate: dueDate, order: order)
    }

    static func aad(accountDid: String, taskId: String, eventId: String) -> Data {
        Data("\(accountDid)|\(taskId)|\(eventId)".utf8)
    }

    /// Build and seal an upsert/delete event for the given task snapshot. The plaintext
    /// for delete events is `{}` per §3.
    static func makeEvent(
        op: String,
        accountDid: String,
        deviceId: String,
        eventId: String = UUID().uuidString.lowercased(),
        taskId: String,
        updatedAt: Date,
        plaintext: EventPlaintext?,
        encryptionKey: SymmetricKey,
        nonce: Data? = nil
    ) throws -> SyncEvent {
        let pt: Data
        if op == "delete" {
            pt = Data("{}".utf8)
        } else if let plaintext {
            pt = encodePlaintext(plaintext)
        } else {
            pt = Data("{}".utf8)
        }
        let aad = aad(accountDid: accountDid, taskId: taskId, eventId: eventId)
        let (ct, n) = try SymmetricBox.seal(plaintext: pt, key: encryptionKey, aad: aad, nonce: nonce)
        return SyncEvent(
            accountDid: accountDid,
            deviceId: deviceId,
            eventId: eventId,
            taskId: taskId,
            updatedAt: ISO8601.format(updatedAt),
            op: op,
            ciphertext: Base64Url.encode(ct),
            nonce: Base64Url.encode(n)
        )
    }

    /// Verifies the AAD, decrypts the ciphertext, and (for upsert events) returns the
    /// decoded plaintext. Returns nil for `op="delete"`.
    static func openEvent(_ event: SyncEvent, encryptionKey: SymmetricKey) throws -> EventPlaintext? {
        guard let ct = Base64Url.decode(event.ciphertext), let nonce = Base64Url.decode(event.nonce) else {
            throw SymmetricBoxError.invalidCiphertext
        }
        let aad = aad(accountDid: event.accountDid, taskId: event.taskId, eventId: event.eventId)
        let pt = try SymmetricBox.open(ciphertext: ct, nonce: nonce, key: encryptionKey, aad: aad)
        if event.op == "delete" { return nil }
        return decodePlaintext(pt)
    }

    private static func jsonString(_ s: String) -> String {
        // Escape exactly what the JSON RFC requires; dueDate/createdAt are already
        // ASCII ISO 8601 so the typical hot path is just quotes + the input.
        var out = "\""
        for scalar in s.unicodeScalars {
            switch scalar {
            case "\"": out.append("\\\"")
            case "\\": out.append("\\\\")
            case "\u{08}": out.append("\\b")
            case "\u{0C}": out.append("\\f")
            case "\n": out.append("\\n")
            case "\r": out.append("\\r")
            case "\t": out.append("\\t")
            case _ where scalar.value < 0x20:
                out.append(String(format: "\\u%04x", scalar.value))
            default:
                out.append(Character(scalar))
            }
        }
        out.append("\"")
        return out
    }
}

enum ISO8601 {
    private static let formatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let formatterNoFraction: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func format(_ date: Date) -> String {
        formatter.string(from: date)
    }

    static func parse(_ s: String) -> Date? {
        formatter.date(from: s) ?? formatterNoFraction.date(from: s)
    }
}

File: apps/web/src/sync/syncEvent.ts
// Wire format for one synchronisation event. Mirrors apps/macos/Sources/CornerTasks/Sync/SyncEvent.swift.
// See docs/sync-protocol.md §3 / §4.

import { open as boxOpen, seal as boxSeal } from '../crypto/box';

export interface SyncEvent {
  accountDid: string;
  deviceId: string;
  eventId: string;
  taskId: string;
  updatedAt: string;
  op: 'upsert' | 'delete';
  ciphertext: string;
  nonce: string;
}

export interface EventPlaintext {
  title: string;
  createdAt: Date;
  completedAt: Date | null;
  dueDate: Date | null;
  order: number;
}

const b64urlEncode = (bytes: Uint8Array): string => {
  let s = '';
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s).replaceAll('+', '-').replaceAll('/', '_').replace(/=+$/, '');
};

const b64urlDecode = (s: string): Uint8Array => {
  const pad = s.length % 4 === 0 ? '' : '='.repeat(4 - (s.length % 4));
  const std = s.replaceAll('-', '+').replaceAll('_', '/') + pad;
  const bin = atob(std);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
};

export const isoFormat = (d: Date): string => d.toISOString();
export const isoParse = (s: string): Date | null => {
  const t = Date.parse(s);
  return Number.isFinite(t) ? new Date(t) : null;
};

const jsonString = (s: string): string => JSON.stringify(s);

// Hand-built so the byte sequence matches the macOS encoder exactly:
// title, createdAt, completedAt, dueDate, order. Asserted by docs/crypto-vectors.json.
export const encodePlaintext = (pt: EventPlaintext): Uint8Array => {
  let out = '{';
  out += '"title":' + jsonString(pt.title) + ',';
  out += '"createdAt":' + jsonString(isoFormat(pt.createdAt)) + ',';
  out += '"completedAt":' + (pt.completedAt ? jsonString(isoFormat(pt.completedAt)) : 'null') + ',';
  out += '"dueDate":' + (pt.dueDate ? jsonString(isoFormat(pt.dueDate)) : 'null') + ',';
  out += '"order":' + String(pt.order);
  out += '}';
  return new TextEncoder().encode(out);
};

export const decodePlaintext = (bytes: Uint8Array): EventPlaintext | null => {
  let obj: unknown;
  try {
    obj = JSON.parse(new TextDecoder().decode(bytes));
  } catch {
    return null;
  }
  if (!obj || typeof obj !== 'object') return null;
  const o = obj as Record<string, unknown>;
  if (typeof o.title !== 'string') return null;
  if (typeof o.createdAt !== 'string') return null;
  const createdAt = isoParse(o.createdAt);
  if (!createdAt) return null;
  let completedAt: Date | null = null;
  if (typeof o.completedAt === 'string') completedAt = isoParse(o.completedAt);
  let dueDate: Date | null = null;
  if (typeof o.dueDate === 'string') dueDate = isoParse(o.dueDate);
  const order = typeof o.order === 'number' ? o.order : 0;
  return { title: o.title, createdAt, completedAt, dueDate, order };
};

export const eventAad = (accountDid: string, taskId: string, eventId: string): Uint8Array =>
  new TextEncoder().encode(`${accountDid}|${taskId}|${eventId}`);

export interface MakeEventArgs {
  op: 'upsert' | 'delete';
  accountDid: string;
  deviceId: string;
  eventId?: string;
  taskId: string;
  updatedAt: Date;
  plaintext: EventPlaintext | null;
  encryptionKey: Uint8Array;
  nonce?: Uint8Array;
}

const newEventId = (): string => {
  if (typeof crypto !== 'undefined' && 'randomUUID' in crypto) return crypto.randomUUID();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
};

export const makeEvent = async (args: MakeEventArgs): Promise<SyncEvent> => {
  const eventId = args.eventId ?? newEventId();
  let pt: Uint8Array;
  if (args.op === 'delete') {
    pt = new TextEncoder().encode('{}');
  } else if (args.plaintext) {
    pt = encodePlaintext(args.plaintext);
  } else {
    pt = new TextEncoder().encode('{}');
  }
  const aad = eventAad(args.accountDid, args.taskId, eventId);
  const sealed = await boxSeal(pt, args.encryptionKey, aad, args.nonce);
  return {
    accountDid: args.accountDid,
    deviceId: args.deviceId,
    eventId,
    taskId: args.taskId,
    updatedAt: isoFormat(args.updatedAt),
    op: args.op,
    ciphertext: b64urlEncode(sealed.ciphertext),
    nonce: b64urlEncode(sealed.nonce),
  };
};

// Returns null for delete events (and on decode failure of an upsert plaintext). Throws on AEAD failure.
export const openEvent = async (event: SyncEvent, encryptionKey: Uint8Array): Promise<EventPlaintext | null> => {
  const ct = b64urlDecode(event.ciphertext);
  const nonce = b64urlDecode(event.nonce);
  const aad = eventAad(event.accountDid, event.taskId, event.eventId);
  const pt = await boxOpen(ct, nonce, encryptionKey, aad);
  if (event.op === 'delete') return null;
  return decodePlaintext(pt);
};
